### ISSUE TITLE
10 unit tests in Checkstyle master branch fail when executed in non-English locale

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#3377]
Checkstyle's UTs have problems being executed on German machine (https://github.com/checkstyle/checkstyle/issues/3317). In addition, Maven does not set user language and country properly when it executes UTs with the following set of commands taken from Travis CI matrix:

```
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=de -Duser.country=DE'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=es -Duser.country=ES'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=fi -Duser.country=FI'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=fr -Duser.country=FR'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=zh -Duser.country=CN'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=ja -Duser.country=JP'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=pt -Duser.country=PT'
mvn clean integration-test failsafe:verify -DargLine='-Duser.language=tr -Duser.country=TR'
```

BaseCheckTestSupport has [getCheckMessage](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java#L265) method which loads message resources as stream, but not as resource bundle. It means that UTs have to fail in German or other locale as we always load the messages.properties without language and country variants, but they do not.

The problem may happen due to the fact that some system properties (as `user.language` and `user.country`) must be set on the command line of the JVM, and cannot be set after the JVM has been started. 

maven-surefire-plugin has [special VM properties](http://maven.apache.org/surefire/maven-surefire-plugin/examples/system-properties.html) which are intended to deal with the Maven's problem. If we configure maven surefire plugin in our pom.xml as follows:

``` xml
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>2.19.1</version>
        <configuration>
          <systemPropertyVariables>
            <jacoco-agent.destfile>${project.build.directory}/jacoco/jacoco.exec</jacoco-agent.destfile>
          </systemPropertyVariables>
          <additionalClasspathElements>
            <additionalClasspathElement>src/test/resources-noncompilable</additionalClasspathElement>
          </additionalClasspathElements>
          <includes>
            <include>com/puppycrawl/**/*.java</include>
          </includes>
          <argLine>-Duser.country=DE -Duser.language=de</argLine>
        </configuration>
      </plugin>
```

and run Maven from command line, there will be 10 failed UTs:

```
Failed tests:
  CheckerTest.testMultipleConfigs:633->BaseCheckTestSupport.verify:126->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...963750\file.java:1: [Missing a header - not enough lines in file].> but was:<...963750\file.java:1: [Header fehlt - Datei hat nicht genug Zeilen].>
  HeaderCheckTest.testCacheHeaderFile:266->BaseCheckTestSupport.verify:126->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...InputHeader.java:1: [Missing a header - not enough lines in file].> but was:<...InputHeader.java:1: [Header fehlt - Datei hat nicht genug Zeilen].>
  HeaderCheckTest.testNotMatch:185->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...igs\java2.header:2: [Line does not match expected header line of] '// checkstyle: Che...> but was:<...igs\java2.header:2: [Zeile entspricht nicht der erwarteten Header-Zeile] '// checkstyle: Che...>
  HeaderCheckTest.testStaticHeader:75->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...InputHeader.java:1: [Missing a header - not enough lines in file].> but was:<...InputHeader.java:1: [Header fehlt - Datei hat nicht genug Zeilen].>
  RegexpHeaderCheckTest.testEmptyMultiline:273->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...SmallHeader.java:1: [Missing a header - not enough lines in file].> but was:<...SmallHeader.java:1: [Header fehlt - Datei hat nicht genug Zeilen].>
  RegexpHeaderCheckTest.testInlineRegexpHeader:162->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...gexpHeader7.java:3: [Line does not match expected header line of] '// Created: 2002'.> but was:<...gexpHeader7.java:3: [Zeile entspricht nicht der erwarteten Header-Zeile] '// Created: 2002'.>
  RegexpHeaderCheckTest.testRegexpHeader:139->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...gexpHeader7.java:3: [Line does not match expected header line of] '// Created: 2002'.> but was:<...gexpHeader7.java:3: [Zeile entspricht nicht der erwarteten Header-Zeile] '// Created: 2002'.>
  RegexpHeaderCheckTest.testRegexpHeaderMulti5:241->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...gexpHeader4.java:1: [Missing a header - not enough lines in file].> but was:<...gexpHeader4.java:1: [Header fehlt - Datei hat nicht genug Zeilen].>
  RegexpHeaderCheckTest.testRegexpHeaderMulti52:284->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...gexpHeader5.java:1: [Missing a header - not enough lines in file].> but was:<...gexpHeader5.java:1: [Header fehlt - Datei hat nicht genug Zeilen].>
  RegexpHeaderCheckTest.testRegexpHeaderUrl:151->BaseCheckTestSupport.verify:121->BaseCheckTestSupport.verify:134->BaseCheckTestSupport.verify:162 error message 0 expected:<...gexpHeader7.java:3: [Line does not match expected header line of] '// Created: 2002'.> but was:<...gexpHeader7.java:3: [Zeile entspricht nicht der erwarteten Header-Zeile] '// Created: 2002'.>
```

As you can see expected message was in English but actual was in German. Note, that the problem can be reproduced only if all JVM instances are terminated before Maven execution. The main problem is that we do not know why tests which do not extend AbstractFileSetCheck directly are not broken.
##

**Comment by MEZk** [comment]
Unstable UTs in master
https://travis-ci.org/checkstyle/checkstyle/jobs/151702012#L801

**Comment by romani** [comment]
on my pc (but some time it pass so failure is unstable):

```
$ mvn clean integration-test failsafe:verify -DargLine='-Duser.language=de -Duser.country=DE'
....................
Failed tests: 
  HeaderCheckTest.testCacheHeaderFile:265->BaseCheckTestSupport.verify:211->BaseCheckTestSupport.verify:231->BaseCheckTestSupport.verify:262 error message 0 expected:<...InputHeader.java:1: [Missing a header - not enough lines in file].> but was:<...InputHeader.java:1: [Die erforderliche Kopfzeile fehlt - Datei hat nicht genug Zeilen].>
  HeaderCheckTest.testCacheHeaderWithoutFile:289->BaseCheckTestSupport.verify:211->BaseCheckTestSupport.verify:231->BaseCheckTestSupport.verify:262 error message 0 expected:<...InputHeader.java:1: [Line does not match expected header line of] 'Test'.> but was:<...InputHeader.java:1: [Zeile entspricht nicht der erwarteten Kopfzeile] 'Test'.>
  HeaderCheckTest.testNotMatch:185->BaseCheckTestSupport.verify:195->BaseCheckTestSupport.verify:231->BaseCheckTestSupport.verify:262 error message 0 expected:<...igs/java2.header:2: [Line does not match expected header line of] '// checkstyle: Che...> but was:<...igs/java2.header:2: [Zeile entspricht nicht der erwarteten Kopfzeile] '// checkstyle: Che...>
  HeaderCheckTest.testStaticHeader:74->BaseCheckTestSupport.verify:195->BaseCheckTestSupport.verify:231->BaseCheckTestSupport.verify:262 error message 0 expected:<...InputHeader.java:1: [Missing a header - not enough lines in file].> but was:<...InputHeader.java:1: [Die erforderliche Kopfzeile fehlt - Datei hat nicht genug Zeilen].>

Tests run: 1977, Failures: 4, Errors: 0, Skipped: 0

```

**Comment by Vladlis** [comment]
May we try to move these tests to another CI?

**Comment by romani** [comment]
I closing this issue as it suppose to be fixed by #3896 .

If smb still experience this problems , please reopen issue.
